### PR TITLE
Improve and fix fake_quantize_op

### PR DESCRIPTION
**Pull request by qingqing01**
- Improve the original code.
- Fix bug in CPU and CUDA kernel.

### CODE REVIEW

**Review by wanghaoshuang — APPROVED** [review]
LGTM

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 1]
```diff
@@ -14,86 +14,198 @@ limitations under the License. */
...
+    int64_t it = iter.data<int64_t>()[0];
+    int idx = it % window_size;
+    T removed = scale_arr[idx];
+    T cur = cur_scale.data<T>()[0];
+    scale_arr[idx] = cur;
+
+    T max = last_scale.data<T>()[0];
+    if (max < cur) {
+      max = cur;
+    } else if (fabs(removed - max) < 1e-6) {
```
Comment (panyx0718) [inline paddle/fluid/operators/fake_quantize_op.cc:84]: don't quite understand this part.

File: paddle/fluid/operators/fake_quantize_op.cu [hunk 2]
```diff
@@ -53,220 +54,124 @@ __global__ void FindAbsMaxKernel(const int n, const T* in, T* out) {
...
+template struct FindAbsMaxFunctor<platform::CUDADeviceContext, float>;
 
 template <typename T>
-__global__ void ApplySaturateKernel(const int n, const T* in, T* out,
-                                    int* num_saturate, const T min,
-                                    const T max) {
+__global__ void ClipAndQuantKernel(const T* in, const T* scale,
+                                   const int bin_cnt, const int n, T* out) {
   int bid = threadIdx.x + blockIdx.x * blockDim.x;
   int tid = threadIdx.x;
```
Comment (panyx0718) [inline paddle/fluid/operators/fake_quantize_op.cu:81]: not used?

File: paddle/fluid/operators/fake_quantize_op.cc [hunk 3]
```diff
@@ -14,86 +14,198 @@ limitations under the License. */
...
+        .SetDefault(10000);
+    AddAttr<int>("bit_length", "(int, default 8), quantization bit number.")
+        .SetDefault(8)
+        .AddCustomChecker([](const int& bit_length) {
+          PADDLE_ENFORCE(bit_length >= 1 && bit_length <= 16,
+                         "'bit_length' should be between 1 and 16.");
+        });
+    AddAttr<bool>("is_test", "").SetDefault(false);
+    AddComment(R"DOC(
+FakeQuantize operator is used in static quantization.
```
Comment (panyx0718) [inline paddle/fluid/operators/fake_quantize_op.cc:204]: Is this comment correct? Maybe a bit more detail?

**Pull request merged**